Cleanup of report todos and spelling (#93)
chhoumann/p7: report/sections/Chapter 05 - Architecture/Architecture.tex

In this chapter we will describe the architecture of the entire system by first showing the overall structure followed by a short description of the individual components. 
The platform allows lecturers to create a syllabus that includes problem sets, exercises and test cases for the exercises. Additionally, students can do the exercises that the lecturer has provided. 
The platform is a web application designed using a client-server architecture. Using this architecture the client communicates with a server using HTTP to send \textit{request} messages. The server waits for client requests and processes the requests when they arrive and replies to the requests with a \textit{response} message.

> The platform has a user-interface as well as business logic to handle user requests and authentication with roles, which means users of varying roles have different permissions. The platform can run tests against code, to check whether it satisfies given requirements. Account data, test runs, and syllabus data is persisted in a database.

Reviewer: ```suggestion
The platform has a user-interface as well as business logic to handle user requests. The platform supports role-based authentication with roles, meaning that users of varying roles have access to different functionality. The platform can run tests, to check whether a piece of code solves an exercise. Account data, test runs, and syllabus data is persisted in a database.
```

Reviewer: Vi har defineret hvad en test er, ingen grund til at sige at vi hvad en test er igen.

	\label{fig:Architecture}
\end{figure}

> Note that the platform also has an API running on a separate server instance to handle code execution and test, called \textit{Test runner}. As each component is introduced, in the following sections, we will elaborate on the choices made for each component, in relation to this structure.

Reviewer: ```suggestion
Note that the platform has an API running on a separate server instance to handle code execution and test, called \textit{Test runner}. As components are introduced in the following sections, we elaborate on the choices made for each component, in relation to this structure.
```

In order to support different roles in the system, we have created the entities \textit{Role} and \textit{User}. The \textit{Role} entity has one attribute called \textbf{Name}. This corresponds to the name of the role. The \textit{User} entity has the attributes \textbf{Id}, which is a unique identifier, \textbf{Name}, \textbf{Email}, \textbf{EmailVerifed} which is when the email was verified and an \textbf{Image} attribute which contains the URL for an image of the user. The \textit{User} entity has a many-to-one relationship with the entity \textit{Role}, which means that users can only have one role.

To support a lecturer in creating a syllabus for a semester course we have created the entities \textit{Syllabus}, \textit{Problemset}, \textit{Problem} and \textit{test}. For each entity, a set of attributes have been added to contain the relevant data and relationships between these entities to annotate their relation.

> For the \textit{Problem} entity we have added an \textbf{Id} attribute as a unique identifier, a \textbf{Name} attribute, a \textbf{Template} attribute, which contains code that will be automatically included in the editor when starting a problem. This ensures that the functions required for the tests are included, as described in the \textbf{Hspec} section of chapter \ref{chap:preliminaries}. The final attribute is the \textbf{Description}, which contains the description of the exercise.

Reviewer: ```suggestion
For the \textit{Problem} entity we have added an \textbf{Id} attribute as a unique identifier, a \textbf{Name} attribute, a \textbf{Template} attribute, which contains code that will be automatically included in the editor when starting a problem. This ensures that the functions  are encluded in a module which can later be accessed by \textbf{Hspec}. The final attribute is the \textbf{Description}, which contains the description of the exercise.
```

Reviewer 2: ```suggestion
For the \textit{Problem} entity we have added an \textbf{Id} attribute as a unique identifier, a \textbf{Name} attribute, a \textbf{Template} attribute, which contains code that will be automatically included in the editor when starting a problem. This ensures that the functions are included in a module which can later be accessed by \textbf{Hspec}. The final attribute is the \textbf{Description}, which contains the description of the exercise.```
